File: database/operations.py
from .models import User, Watchlist, WatchlistToken, Transaction, CopyTradeConfig, get_db_session
from sqlalchemy.exc import IntegrityError
from sqlalchemy import and_, or_
import sys
import os

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# User operations
def create_user(telegram_id, username=None, first_name=None):
    """Create a new user in the database"""
    db = get_db_session()
    try:
        user = User(
            telegram_id=str(telegram_id),
            username=username,
            first_name=first_name
        )
        db.add(user)
        db.commit()
        db.refresh(user)
        return user
    except IntegrityError:
        db.rollback()
        return get_user_by_telegram_id(telegram_id)
    finally:
        db.close()

def get_user_by_telegram_id(telegram_id):
    """Get a user by Telegram ID"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        return user
    finally:
        db.close()

def update_user_wallet(telegram_id, wallet_address, encrypted_private_key):
    """Update a user's wallet information"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if user:
            user.wallet_address = wallet_address
            user.encrypted_private_key = encrypted_private_key
            db.commit()
            db.refresh(user)
            return user
        return None
    finally:
        db.close()

def update_user_settings(telegram_id, slippage=None, auto_buy=None, auto_buy_amount=None):
    """Update a user's settings"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if user:
            if slippage is not None:
                user.slippage = slippage
            if auto_buy is not None:
                user.auto_buy = auto_buy
            if auto_buy_amount is not None:
                user.auto_buy_amount = auto_buy_amount
            db.commit()
            db.refresh(user)
            return user
        return None
    finally:
        db.close()

# Watchlist operations
def create_watchlist(telegram_id, name):
    """Create a new watchlist for a user"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if not user:
            return None
        
        watchlist = Watchlist(
            user_id=user.id,
            name=name
        )
        db.add(watchlist)
        db.commit()
        db.refresh(watchlist)
        return watchlist
    finally:
        db.close()

def get_user_watchlists(telegram_id):
    """Get all watchlists for a user"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if not user:
            return []
        
        return db.query(Watchlist).filter(Watchlist.user_id == user.id).all()
    finally:
        db.close()

def add_token_to_watchlist(watchlist_id, token_address, token_symbol=None, alert_price_high=None, alert_price_low=None):
    """Add a token to a watchlist"""
    db = get_db_session()
    try:
        watchlist_token = WatchlistToken(
            watchlist_id=watchlist_id,
            token_address=token_address,
            token_symbol=token_symbol,
            alert_price_high=alert_price_high,
            alert_price_low=alert_price_low
        )
        db.add(watchlist_token)
        db.commit()
        db.refresh(watchlist_token)
        return watchlist_token
    finally:
        db.close()

def get_watchlist_tokens(watchlist_id):
    """Get all tokens in a watchlist"""
    db = get_db_session()
    try:
        return db.query(WatchlistToken).filter(WatchlistToken.watchlist_id == watchlist_id).all()
    finally:
        db.close()

# Transaction operations
def create_transaction(telegram_id, transaction_hash, token_address, token_symbol, amount, eth_amount, transaction_type):
    """Create a new transaction record"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if not user:
            return None
        
        transaction = Transaction(
            user_id=user.id,
            transaction_hash=transaction_hash,
            token_address=token_address,
            token_symbol=token_symbol,
            amount=amount,
            eth_amount=eth_amount,
            transaction_type=transaction_type
        )
        db.add(transaction)
        db.commit()
        db.refresh(transaction)
        return transaction
    finally:
        db.close()

def update_transaction_status(transaction_hash, status):
    """Update a transaction's status"""
    db = get_db_session()
    try:
        transaction = db.query(Transaction).filter(Transaction.transaction_hash == transaction_hash).first()
        if transaction:
            transaction.status = status
            db.commit()
            db.refresh(transaction)
            return transaction
        return None
    finally:
        db.close()

def get_user_transactions(telegram_id, limit=10):
    """Get a user's transactions"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if not user:
            return []
        
        return db.query(Transaction).filter(Transaction.user_id == user.id).order_by(Transaction.created_at.desc()).limit(limit).all()
    finally:
        db.close()

# Copy trade operations
def create_copy_trade_config(telegram_id, target_wallet, slippage_guard=5.0, active=True, sandbox_mode=False, max_eth_per_trade=0.1):
    """Create a new copy trade configuration"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if not user:
            return None
        
        config = CopyTradeConfig(
            user_id=user.id,
            target_wallet=target_wallet,
            slippage_guard=slippage_guard,
            active=active,
            sandbox_mode=sandbox_mode,
            max_eth_per_trade=max_eth_per_trade
        )
        db.add(config)
        db.commit()
        db.refresh(config)
        return config
    finally:
        db.close()

def update_copy_trade_config(config_id, active=None, slippage_guard=None, sandbox_mode=None, max_eth_per_trade=None):
    """Update a copy trade configuration"""
    db = get_db_session()
    try:
        config = db.query(CopyTradeConfig).filter(CopyTradeConfig.id == config_id).first()
        if config:
            if active is not None:
                config.active = active
            if slippage_guard is not None:
                config.slippage_guard = slippage_guard
            if sandbox_mode is not None:
                config.sandbox_mode = sandbox_mode
            if max_eth_per_trade is not None:
                config.max_eth_per_trade = max_eth_per_trade
            db.commit()
            db.refresh(config)
            return config
        return None
    finally:
        db.close()

def get_user_copy_trade_configs(telegram_id):
    """Get a user's copy trade configurations"""
    db = get_db_session()
    try:
        user = db.query(User).filter(User.telegram_id == str(telegram_id)).first()
        if not user:
            return []
        
        return db.query(CopyTradeConfig).filter(CopyTradeConfig.user_id == user.id).all()
    finally:
        db.close() 

File: database/operations.js
/**
 * Database Operations for Zoracle Telegram Bot
 */
const { User, Transaction, Alert, CopyTrade, Token } = require('./models');
const { Op } = require('sequelize');

/**
 * User Operations
 */
const UserOps = {
  /**
   * Create or update a user
   * @param {string} telegramId - Telegram user ID
   * @param {Object} userData - User data to update
   * @returns {Promise<Object>} - Updated user object
   */
  async upsertUser(telegramId, userData) {
    const [user, created] = await User.findOrCreate({
      where: { telegramId },
      defaults: userData
    });
    
    if (!created && userData) {
      await user.update(userData);
    }
    
    return user.toJSON();
  },
  
  /**
   * Get a user by Telegram ID
   * @param {string} telegramId - Telegram user ID
   * @returns {Promise<Object|null>} - User object or null if not found
   */
  async getUser(telegramId) {
    const user = await User.findByPk(telegramId);
    return user ? user.toJSON() : null;
  },
  
  /**
   * Update user's last active timestamp
   * @param {string} telegramId - Telegram user ID
   * @returns {Promise<boolean>} - Success status
   */
  async updateLastActive(telegramId) {
    const result = await User.update(
      { lastActive: new Date() },
      { where: { telegramId } }
    );
    return result[0] > 0;
  }
};

/**
 * Transaction Operations
 */
const TransactionOps = {
  /**
   * Create a new transaction
   * @param {Object} txData - Transaction data
   * @returns {Promise<Object>} - Created transaction
   */
  async createTransaction(txData) {
    const tx = await Transaction.create(txData);
    return tx.toJSON();
  },
  
  /**
   * Update a transaction
   * @param {string} id - Transaction ID
   * @param {Object} txData - Transaction data to update
   * @returns {Promise<boolean>} - Success status
   */
  async updateTransaction(id, txData) {
    const result = await Transaction.update(
      txData,
      { where: { id } }
    );
    return result[0] > 0;
  },
  
  /**
   * Update a transaction by tx hash
   * @param {string} txHash - Transaction hash
   * @param {Object} txData - Transaction data to update
   * @returns {Promise<boolean>} - Success status
   */
  async updateTransactionByHash(txHash, txData) {
    const result = await Transaction.update(
      txData,
      { where: { txHash } }
    );
    return result[0] > 0;
  },
  
  /**
   * Get user's transactions
   * @param {string} telegramId - Telegram user ID
   * @param {number} limit - Number of transactions to return
   * @param {number} offset - Offset for pagination
   * @returns {Promise<Array>} - Array of transactions
   */
  async getUserTransactions(telegramId, limit = 10, offset = 0) {
    const txs = await Transaction.findAll({
      where: { telegramId },
      order: [['timestamp', 'DESC']],
      limit,
      offset
    });
    return txs.map(tx => tx.toJSON());
  },
  
  /**
   * Get user's transactions for a specific token
   * @param {string} telegramId - Telegram user ID
   * @param {string} tokenAddress - Token address
   * @returns {Promise<Array>} - Array of transactions
   */
  async getUserTokenTransactions(telegramId, tokenAddress) {
    const txs = await Transaction.findAll({
      where: { 
        telegramId,
        tokenAddress
      },
      order: [['timestamp', 'DESC']]
    });
    return txs.map(tx => tx.toJSON());
  }
};

/**
 * Alert Operations
 */
const AlertOps = {
  /**
   * Create a new alert
   * @param {Object} alertData - Alert data
   * @returns {Promise<Object>} - Created alert
   */
  async createAlert(alertData) {
    const alert = await Alert.create(alertData);
    return alert.toJSON();
  },
  
  /**
   * Get user's alerts
   * @param {string} telegramId - Telegram user ID
   * @returns {Promise<Array>} - Array of alerts
   */
  async getUserAlerts(telegramId) {
    const alerts = await Alert.findAll({
      where: { 
        telegramId,
        active: true
      },
      order: [['createdAt', 'DESC']]
    });
    return alerts.map(alert => alert.toJSON());
  },
  
  /**
   * Get active alerts for a specific token
   * @param {string} tokenAddress - Token address
   * @returns {Promise<Array>} - Array of alerts
   */
  async getTokenAlerts(tokenAddress) {
    const alerts = await Alert.findAll({
      where: { 
        tokenAddress,
        active: true
      }
    });
    return alerts.map(alert => alert.toJSON());
  },
  
  /**
   * Deactivate an alert
   * @param {string} id - Alert ID
   * @returns {Promise<boolean>} - Success status
   */
  async deactivateAlert(id) {
    const result = await Alert.update(
      { active: false },
      { where: { id } }
    );
    return result[0] > 0;
  },
  
  /**
   * Update alert's last triggered timestamp
   * @param {string} id - Alert ID
   * @returns {Promise<boolean>} - Success status
   */
  async updateAlertTriggered(id) {
    const result = await Alert.update(
      { lastTriggered: new Date() },
      { where: { id } }
    );
    return result[0] > 0;
  }
};

/**
 * CopyTrade Operations
 */
const CopyTradeOps = {
  /**
   * Create a new copy trade configuration
   * @param {Object} copyTradeData - Copy trade data
   * @returns {Promise<Object>} - Created copy trade configuration
   */
  async createCopyTrade(copyTradeData) {
    const copyTrade = await CopyTrade.create(copyTradeData);
    return copyTrade.toJSON();
  },
  
  /**
   * Get user's copy trade configurations
   * @param {string} telegramId - Telegram user ID
   * @returns {Promise<Array>} - Array of copy trade configurations
   */
  async getUserCopyTrades(telegramId) {
    const copyTrades = await CopyTrade.findAll({
      where: { telegramId },
      order: [['createdAt', 'DESC']]
    });
    return copyTrades.map(ct => ct.toJSON());
  },
  
  /**
   * Get all active copy trades for a target wallet
   * @param {string} targetWallet - Target wallet address
   * @returns {Promise<Array>} - Array of copy trade configurations
   */
  async getActiveTargetCopyTrades(targetWallet) {
    const copyTrades = await CopyTrade.findAll({
      where: { 
        targetWallet,
        active: true
      }
    });
    return copyTrades.map(ct => ct.toJSON());
  },
  
  /**
   * Update a copy trade configuration
   * @param {string} id - Copy trade ID
   * @param {Object} copyTradeData - Copy trade data to update
   * @returns {Promise<boolean>} - Success status
   */
  async updateCopyTrade(id, copyTradeData) {
    const result = await CopyTrade.update(
      copyTradeData,
      { where: { id } }
    );
    return result[0] > 0;
  },
  
  /**
   * Toggle copy trade active status
   * @param {string} id - Copy trade ID
   * @param {boolean} active - Active status
   * @returns {Promise<boolean>} - Success status
   */
  async toggleCopyTradeActive(id, active) {
    const result = await CopyTrade.update(
      { active },
      { where: { id } }
    );
    return result[0] > 0;
  }
};

/**
 * Token Operations
 */
const TokenOps = {
  /**
   * Upsert token information
   * @param {string} address - Token address
   * @param {Object} tokenData - Token data
   * @returns {Promise<Object>} - Updated token object
   */
  async upsertToken(address, tokenData) {
    const [token, created] = await Token.findOrCreate({
      where: { address },
      defaults: { address, ...tokenData }
    });
    
    if (!created && tokenData) {
      await token.update(tokenData);
    }
    
    return token.toJSON();
  },
  
  /**
   * Get token information
   * @param {string} address - Token address
   * @returns {Promise<Object|null>} - Token object or null if not found
   */
  async getToken(address) {
    const token = await Token.findByPk(address);
    return token ? token.toJSON() : null;
  },
  
  /**
   * Get all Zora tokens
   * @param {number} limit - Number of tokens to return
   * @param {number} offset - Offset for pagination
   * @returns {Promise<Array>} - Array of tokens
   */
  async getZoraTokens(limit = 20, offset = 0) {
    const tokens = await Token.findAll({
      where: { isZoraToken: true },
      order: [['priceUpdateTime', 'DESC']],
      limit,
      offset
    });
    return tokens.map(token => token.toJSON());
  },
  
  /**
   * Search tokens by name or symbol
   * @param {string} query - Search query
   * @param {number} limit - Number of tokens to return
   * @returns {Promise<Array>} - Array of tokens
   */
  async searchTokens(query, limit = 10) {
    const tokens = await Token.findAll({
      where: {
        [Op.or]: [
          { name: { [Op.like]: `%${query}%` } },
          { symbol: { [Op.like]: `%${query}%` } }
        ]
      },
      limit
    });
    return tokens.map(token => token.toJSON());
  }
};

module.exports = {
  UserOps,
  TransactionOps,
  AlertOps,
  CopyTradeOps,
  TokenOps
}; 